File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        
        leftOrRight = {'(': 'Left', '{': 'Left', '[' : 'Left',
              ')': 'Right', '}': 'Right', ']' : 'Right'}
        
        stack = []
        
        
        for c in s:
            if leftOrRight[c] == 'Left':
                stack.append(c)
            else:
                if not stack:
                    return False
                lastLeft = stack.pop()
                if not isMatching(lastLeft, c):
                    return False
        
        if stack:
            return False
        
        return True

def isMatching(left, right):
    if (left == '(' and right == ')') or (left == '{' and right == '}') or (left == '[' and right == ']'):
        return True
    return False
                
sol = Solution()
sol.isValid('[({})]')               
        
        
        

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isLeft(char c)
    {
        if (c == '(' || c == '{' || c == '[') return true;
        return false;
    }
    
    bool isMatch(char right, char left) {
        if((left == '(' && right == ')') || (left == '{' && right == '}') || (left == '[' && right == ']')) return true;
        return false;
    }
    
    bool isValid(string s) {
        stack<char> stack;
        int size = s.length();
        char cur,popped;
        for (int i = 0; i < size; i++)
        {
            cur = s[i];
            if (isLeft(cur)) stack.push(cur);
            else {
                if (stack.empty()) {
                    return false;
                }
                popped = stack.top();
                stack.pop();
                if (!isMatch(cur, popped)) return false;
            }
        }
        if (stack.empty()) return true;
        return false;



    }
};